java.hints/src/org/netbeans/modules/java/hints/errors/ConvertInvalidVarToExplicitArrayType.java

```diff
@@ -0,0 +1,197 @@
+})
+public class ConvertInvalidVarToExplicitArrayType implements ErrorRule<Void> {
+
+    private static final Set<String> CODES;
+    private static final Map<TypeKind, Integer> NUMERIC_PRIMITIVE_TYPE_PRIORITY_MAP;
+
+    static {
+        Set<String> codes = new HashSet<>();
+        codes.add("compiler.err.cant.infer.local.var.type"); // NOI18N
+        CODES = Collections.unmodifiableSet(codes);
+
+        Map<TypeKind, Integer> map = new HashMap<>();
+        map.put(TypeKind.BYTE, 1);
+        map.put(TypeKind.SHORT, 2);
+        map.put(TypeKind.INT, 3);
+        map.put(TypeKind.LONG, 4);
+        map.put(TypeKind.FLOAT, 5);
+        map.put(TypeKind.DOUBLE, 6);
+        NUMERIC_PRIMITIVE_TYPE_PRIORITY_MAP = Collections.unmodifiableMap(map);
+    }
+
+    @Override
+    public Set<String> getCodes() {
+        return CODES;
+    }
+
+    @Override
+    public List<Fix> run(CompilationInfo compilationInfo, String diagnosticKey, int offset, TreePath treePath, Data<Void> data) {
+
+        if (treePath.getParentPath() == null) {
+            return null;
+        }
+
+        TypeMirror arrayTypeMirror = null;
+        if (treePath.getLeaf().getKind() == Tree.Kind.VARIABLE) {
+            VariableTree oldVariableTree = (VariableTree) treePath.getLeaf();
+            NewArrayTree arrayTree = (NewArrayTree) oldVariableTree.getInitializer();
+            List<? extends ExpressionTree> currentValues = arrayTree.getInitializers();
+
+            TypeMirror arrayElementTypeMirror = null;
+            Integer maxTypePriority = -1;
+
+            boolean isHomoGeneousArray = true;
+            for (ExpressionTree tree : currentValues) {
+
+                Scope s = compilationInfo.getTrees().getScope(treePath);
+
+                arrayElementTypeMirror = compilationInfo.getTreeUtilities().attributeTree(tree, s);
+
+                if (tree instanceof NewClassTree) {
+                    NewClassTree nct = (NewClassTree) tree;
+                    if (nct.getIdentifier().getKind() == Tree.Kind.PARAMETERIZED_TYPE) {
+
+                        return null;
+                    }
+                }
+
+                if (arrayTypeMirror == null) {
+                    arrayTypeMirror = arrayElementTypeMirror;
+                }
+
+                isHomoGeneousArray = compilationInfo.getTypes().isSameType(arrayElementTypeMirror, arrayTypeMirror);
+                if (isHomoGeneousArray) {
+                    continue;
+                }
+
+                // Hint will be enabled only for primitive Numeric array initializers or for homogeneous array members.
+                if (!isHomoGeneousArray && !isPrimitiveNumeric(arrayElementTypeMirror.getKind())) {
+                    return null;
+                }
+
+                if (!isHomoGeneousArray) {
+                    Integer arrayElementPriority = NUMERIC_PRIMITIVE_TYPE_PRIORITY_MAP.get(arrayElementTypeMirror.getKind());
+                    if (maxTypePriority == -1) {
```

Comment:
This check appears to be unnecessary, right? If maxTypePriority == -1, then maxTypePriority < arrayElementPriority, and the next if statement will set the maxTypePriority?

Reply:
Poulation of arrayElementPriority/maxTypePriority starts when it found out array is hetrogeneous.
 In the below scenarion it will detect array is hetrogenous when it traverse 2nd element
{2.1F,1}
it that point
maxTypePriority =-1,arrayElementPriority = 3(int),arrayTypeMirror = float type

If we use only the below check,this will populate maxTypePriority = 3 (int) which is incorrect.
if(maxTypePriority < arrayElementPriority)
 maxTypePriority = arrayElementPriority;

For this reason using the  (maxTypePriority == -1) check

Reply:
Ah, right, sorry I got confused with the names.

I was thinking on how to simplify the logic. And this might work (untested):
TypeMirror arrayType = null;

for (ExpressionTree et : currentValues) {
     TypeMirror etType = ...;
     if (arrayType == null) {
         arrayType = etType;
     } else if (!types.isAssignable(etType, arrayType)) {
         if (types.isAssignable(arrayType, etType)) { //here I am not completely sure if isAssignable is transitive
             arrayType = etType;
         } else {
             return null; //the array is not sufficiently homogeneous.
         }
     }
}

Would this work?